rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/dap/RascalDebugAdapter.java

```diff
@@ -0,0 +1,420 @@
+					ITree result = locateBreakableTree((ITree) child, line);
+
+					if (result != null) {
+						return result;
+					}
+				}
+			}
+		}
+        INode node = IRascalValueFactory.getInstance().node(breakable);
+		if (l.getBeginLine() == line && ProductionAdapter.hasAttribute(TreeAdapter.getProduction(tree), IRascalValueFactory.getInstance().constructor(RascalValueFactory.Attr_Tag, node))) {
+			return tree;
+		}
+
+		return null;
+	}
+
+    @Override
+    public CompletableFuture<Void> attach(Map<String, Object> args) {
+        client.initialized();
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<Void> configurationDone(ConfigurationDoneArguments args) {
+        ProcessEventArguments eventArgs = new ProcessEventArguments();
+        eventArgs.setSystemProcessId((int) ProcessHandle.current().pid());
+        eventArgs.setName(LSPTerminalREPL.class.getProtectionDomain().getCodeSource().getLocation().getPath());
+        eventArgs.setIsLocalProcess(true);
+        eventArgs.setStartMethod(ProcessEventArgumentsStartMethod.ATTACH);
+        client.process(eventArgs);
+
+        ThreadEventArguments thread = new ThreadEventArguments();
+        thread.setThreadId(mainThreadID);
+        thread.setReason(ThreadEventArgumentsReason.STARTED);
+        client.thread(thread);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<ThreadsResponse> threads() {
+        return CompletableFuture.supplyAsync(() -> {
+            ThreadsResponse response = new ThreadsResponse();
+            Thread t = new Thread();
+            t.setId(mainThreadID);
+            t.setName("Main Thread");
+            response.setThreads(new Thread[]{
+                t
+            });
+            return response;
+        });
+    }
+
+    @Override
+    public CompletableFuture<StackTraceResponse> stackTrace(StackTraceArguments args) {
+        if(args.getThreadId() != mainThreadID) {
+            return CompletableFuture.completedFuture(new StackTraceResponse());
+        }
+        return CompletableFuture.supplyAsync(() -> {
+            StackTraceResponse response = new StackTraceResponse();
+            IRascalFrame[] stackFrames = suspendedState.getCurrentStackFrames();
+            response.setTotalFrames(stackFrames.length);
+            StackFrame[] stackFramesResponse = new StackFrame[stackFrames.length];
+            IRascalFrame currentFrame = suspendedState.getCurrentStackFrame();
+            ISourceLocation currentLoc = evaluator.getCurrentPointOfExecution() != null ?
+                evaluator.getCurrentPointOfExecution()
+                : URIUtil.rootLocation("stdin");
+            StackFrame frame = new StackFrame();
+            frame.setId(stackFrames.length-1);
+            frame.setName(currentFrame.getName());
+            frame.setLine(currentLoc.getBeginLine());
+            frame.setColumn(currentLoc.getBeginColumn());
```

Comment:
The `cm` ColumnMaps are also threaded through different parts of the terminal (IDEServices), so I think it would be possible to reuse even an instance. However, it is also doable to start a new ColumnMap for every debug session if that is easier to code.

Reply:
True, but we cannot reuse the same ColumnMaps between debug server and the LSP, as it's seperate JVMs. But indeed, since we are getting the full contents of files anyway, we could build a ColumnMaps for these as well.